File: tasks/python3/getVisMetadata.py
#!/usr/bin/env python

from optparse import OptionParser
import os
import json
import httpx
import asyncio

prog = os.path.basename(__file__)
parser = OptionParser(usage="Usage: %s <config> <output_dir>" % prog)
(options, args) = parser.parse_args()
prog = os.path.basename(__file__)

features_file = args[0]
layer_order_file = args[1]
output_file = args[2]

# These are alias or otherwise layers that don't exist in GIBS
skip_layers = [
  'Land_Water_Map',
  'Land_Mask',
  'World_Database_on_Protected_Areas',
]

# NOTE:  Only using these properties at this time
use_keys = [
  'conceptIds',
  'dataCenter',
  'daynight',
  'orbitTracks',
  'orbitDirection',
  'ongoing',
  'layerPeriod',
  'title',
  'subtitle'
]
layer_metadata = {}
daacMap = {}
failed_requests = []

def get_daac(metadata):
  if metadata.get('conceptIds', None) is None:
    return metadata
  for collection in metadata.get("conceptIds"):
    origDataCenter = collection.get("dataCenter", None)
    dataCenter = daacMap.get(origDataCenter, None)
    collection.pop('dataCenter', None)
    if dataCenter is  None:
      continue
    if metadata.get("dataCenter", None) is None:
      metadata["dataCenter"] = [dataCenter]
    elif dataCenter not in metadata["dataCenter"]:
      metadata["dataCenter"].append(dataCenter)
  return metadata

async def get_metadata(client, layer_id, base_url):
  try:
    response = await client.get(base_url + layer_id + '.json')
    metadata = response.json()
    layer_metadata[layer_id] = get_daac(metadata)

    # Remove any props we don't expect to use
    metadata_keys = dict(layer_metadata[layer_id]).keys()
    for key in metadata_keys:
      if key not in use_keys:
        layer_metadata[layer_id].pop(key, None)

  except Exception as e:
    global failed_requests
    if layer_id not in failed_requests:
      failed_requests.append(layer_id)
      print('%s WARNING: Failed to retrieve metadata config for [%s], will retry...' % (prog, layer_id))
    else:
      print('%s WARNING: Failed to retrieve metadata config for [%s] on second attempt' % (prog, layer_id))

async def main(url):
  global skip_layers

  with open(layer_order_file, 'rt', encoding="utf-8") as layer_order:
    layer_ids = json.load(layer_order).get('layerOrder')
    filtered_ids = list(filter(lambda l: l not in skip_layers, layer_ids))
    print('%s: Pulling vis metadata for %s layers... ' % (prog, len(filtered_ids)))

  limits = httpx.Limits(max_keepalive_connections=10, max_connections=10)
  async with httpx.AsyncClient(limits=limits) as client:
    await asyncio.gather(*[get_metadata(client, layer_id, url) for layer_id in filtered_ids])
    await asyncio.gather(*[get_metadata(client, layer_id, url) for layer_id in failed_requests])

    with open(output_file, "w", encoding="utf-8") as fp:
      # Format of this object will determine how this data is combined into wv.json
      json.dump({ 'layers': layer_metadata}, fp, indent=2, sort_keys=True)

#MAIN
if __name__ == "__main__":
  with open(features_file, 'rt', encoding="utf-8") as features:
    metadata_config = json.load(features).get('features').get('vismetadata')
    if metadata_config is not None:
      url = metadata_config.get('url')
      daacMap = metadata_config.get('daacMap', {})
      asyncio.run(main(url))
    else:
      print('%s: Visualization metadata not configured. Exiting.' % (prog))

File: tasks/node/getVisMetadata.js
const fs = require('fs')
const path = require('path')
const yargs = require('yargs')
const console = require('console')
const request = require('axios').default

const prog = path.basename(__filename)

const options = yargs
  .usage('Usage: $0 [options]')
  .option('features', {
    demandOption: true,
    alias: 'f',
    type: 'string',
    description: 'features.json file'
  })
  .option('layerOrder', {
    demandOption: true,
    alias: 'o',
    type: 'string',
    description: 'layerOrder.json file'
  })
  .option('layerMetadata', {
    demandOption: true,
    alias: 'm',
    type: 'string',
    description: 'layer-metadata/all.json file'
  })
  .epilog('Pulls visualization metadata files')

const { argv } = options
if (!argv.features && !argv.layerOrder && !argv.layerMetadata) {
  throw new Error('Invalid number of arguments')
}

const featuresFile = argv.features
let featuresData = fs.readFileSync(featuresFile)
let features = JSON.parse(featuresData)
if (fs.existsSync(featuresFile)) {
  featuresData = fs.readFileSync(featuresFile)
  features = JSON.parse(featuresData)
} else {
  throw new Error(`Error: ${featuresFile} feature file does not exist`)
}

const layerOrderFile = argv.layerOrder
let layerOrderData
let layerOrder
if (fs.existsSync(layerOrderFile)) {
  layerOrderData = fs.readFileSync(layerOrderFile)
  layerOrder = JSON.parse(layerOrderData)
} else {
  throw new Error(`Error: ${layerOrderFile} layer order file does not exist`)
}

const outputFile = argv.layerMetadata
let outputData
let output
if (fs.existsSync(outputFile)) {
  outputData = fs.readFileSync(outputFile)
  output = JSON.parse(outputData)
}

const metadataConfig = features.features.vismetadata
const url = metadataConfig.url
const daacMap = metadataConfig.daacMap || {}
const layerMetadata = {}
const failedRequests = []

// These are alias or otherwise layers that don't exist in GIBS
skipLayers = [
  'Land_Water_Map',
  'Land_Mask',
  'World_Database_on_Protected_Areas'
]

// NOTE:  Only using these properties at this time
useKeys = [
  'conceptIds',
  'dataCenter',
  'daynight',
  'orbitTracks',
  'orbitDirection',
  'ongoing',
  'layerPeriod',
  'title',
  'subtitle'
]

async function main (url) {
  layerOrder = layerOrder.layerOrder
  layerOrder = layerOrder.filter(x => !skipLayers.includes(x))

  console.warn(`${prog}: Pulling vis metadata for ${layerOrder.length} layers... `)
  for (layerId of layerOrder) {
    await getMetadata(layerId, url)
  }
  console.warn(layerMetadata)
}

async function getDAAC (metadata) {
  console.warn(metadata)
}

async function getMetadata (layerId, baseUrl) {
  try {
    await request({
      method: 'get',
      url: `${baseUrl}${layerId}.json`,
      timeout: 10000
    }).then(async (response) => {
      metadata = JSON.parse(response)
      layerMetadata[layerId] = getDAAC(metadata)
    })
  } catch (error) {
    // count how many times this has errored, then rety 2 times
    // throw new Error(`{prog}: WARNING: Failed to retrieve metadata config for ${layerId}, will retry...`)
  }
}

main(url).catch((err) => {
  console.error(err.stack)
  throw new Error(`${prog}: Visualization metadata not configured. Exiting.`)
})
